cpp/arcticdb/util/decimal.cpp

```diff
@@ -0,0 +1,328 @@
+                            current_symbol);
+                    flags_ |= NumberComponentsFlags::DECIMAL;
+                    result.decimal_point_position = i;
+                } else if(is_exponent_symbol(current_symbol)) {
+                    result.exponent_position = i;
+                    break;
+                } else  {
+                    arcticdb::util::check_arg(
+                            is_digit(current_symbol),
+                            "Cannot parse decimal from string. Invalid character '{}'.",
+                            current_symbol);
+                }
+            }
+            return result;
+        }
+
+        void NumberComponents::handle_insignificant_zeros(std::string_view &input) const {
+            if(is_decimal()) {
+                trim_trailing_zeroes(input);
+            }
+        }
+
+        void NumberComponents::parse_significant_digits(std::string_view input) {
+            for(const char symbol : input) {
+                arcticdb::util::check_arg(size_ < max_digits,
+                                          "Cannot parse decimal from string. "
+                                          "Overflow. Input has more than {} significant digits.",
+                                          max_digits);
+                if(ARCTICDB_LIKELY(is_digit(symbol))) {
+                    digits_[size_++] = symbol;
+                } else if(ARCTICDB_UNLIKELY(symbol == '.')) {
+                    continue;
+                }
+            }
+        }
+
+        void NumberComponents::expand_exponent(int decimal_point_position) {
+            arcticdb::util::check_arg(
+                    size_ > 0 || exponent_ == 0,
+                    "Cannot parse decimal from string. "
+                    "There must be at least one digit before the exponent.");
+            const int digits_after_decimal_point = is_decimal() ? size_ - decimal_point_position : 0;
+            const int zeros_to_append = std::max(0, exponent_ - digits_after_decimal_point);
+            arcticdb::util::check_arg(
+                    size_ + zeros_to_append <= max_digits,
+                    "Cannot parse decimal from string. "
+                    "Overflow. Input has more than {} significant digits.",
+                    max_digits);
+            for(int i = 0; i < zeros_to_append; ++i) {
+                digits_[size_++] = '0';
+            }
+        }
+
+        bool NumberComponents::is_negative() const {
+            return flags_ & NumberComponentsFlags::NEGATIVE;
+        }
+
+        bool NumberComponents::is_decimal() const {
+            return flags_ & NumberComponentsFlags::DECIMAL;
+        }
+
+        const char* NumberComponents::get_digits() const {
+            return digits_.data();
+        }
+
+        int NumberComponents::get_size() const {
+            return size_;
+        }
+
+        Decimal::Decimal() : data_{ 0 } {}
+
+        Decimal::Decimal(std::string_view number) : data_{0} {
```

Comment:
I went with custom linear parsing, but regex might be a better idea. Any opinions?